rb_ws/src/buggy/buggy/buggy_state_converter.py

```diff
@@ -0,0 +1,165 @@
+        )
+        self.self_state_publisher = self.create_publisher(Odometry, "self/state", 10)
+
+        # Check if namespace is "/SC" to add an additional "other" subscriber/publisher
+        if namespace == "/SC":
+            self.other_raw_state_subscriber = self.create_subscription(
+                Odometry, "other/raw_state", self.other_raw_state_callback, 10
+            )
+            self.other_state_publisher = self.create_publisher(Odometry, "other/state", 10)
+
+        # Initialize pyproj Transformer for ECEF -> UTM conversion for /SC
+        self.ecef_to_utm_transformer = pyproj.Transformer.from_crs(
+            "epsg:4978", "epsg:32617", always_xy=True
+        )  # Check UTM EPSG code, using EPSG:32617 for UTM Zone 17N
+
+    def self_raw_state_callback(self, msg):
+        """ Callback for processing self/raw_state messages and publishing to self/state """
+        namespace = self.get_namespace()
+
+        if namespace == "/SC":
+            converted_msg = self.convert_SC_state(msg)
+        elif namespace == "/NAND":
+            converted_msg = self.convert_NAND_state(msg)
+        else:
+            self.get_logger().warn("Namespace not recognized for buggy state conversion.")
+            return
+
+        # Publish the converted message to self/state
+        self.self_state_publisher.publish(converted_msg)
+
+    def other_raw_state_callback(self, msg):
+        """ Callback for processing other/raw_state messages and publishing to other/state """
+        # Convert the /SC message and publish to other/state
+        converted_msg = self.convert_SC_state(msg)
+        self.other_state_publisher.publish(converted_msg)
+
+    def convert_SC_state(self, msg):
+        """ Converts /SC namespace raw state to clean state units and structure """
+        converted_msg = Odometry()
+        converted_msg.header = msg.header
+
+        # ---- 1. Convert ECEF Position to UTM Coordinates ----
+        ecef_x = msg.pose.pose.position.x
+        ecef_y = msg.pose.pose.position.y
+        ecef_z = msg.pose.pose.position.z
+
+        # Convert ECEF to UTM
+        utm_x, utm_y, _ = self.ecef_to_utm_transformer.transform(ecef_x, ecef_y, ecef_z)
+        converted_msg.pose.pose.position.x = utm_x  # UTM Easting
+        converted_msg.pose.pose.position.y = utm_y  # UTM Northing
+        converted_msg.pose.pose.position.z = 0.0    # ignored
+
+        # ---- 2. Convert Quaternion to Heading (Radians) ----
+
+        qx, qy, qz, qw = msg.pose.pose.orientation.x, msg.pose.pose.orientation.y, msg.pose.pose.orientation.z, msg.pose.pose.orientation.w
+
+        # replaced with tf_transformations library function
+        # yaw = self.quaternion_to_yaw(qx, qy, qz, qw)
+        # converted_msg.pose.pose.orientation.x = yaw
+
+        # Use euler_from_quaternion to get roll, pitch, yaw
+        _, _, yaw = euler_from_quaternion([qx, qy, qz, qw])
+        yaw_aligned_east = yaw - np.pi / 2   # TODO: check if 0 is East
```

Comment:
quaternion and our internal heading _should_ have the same heading, I wouldn't worry about this.

Reply:
Also, take the other variables and it put into the function